engine/CMakeLists.txt

```diff
@@ -0,0 +1,22 @@
+set(SOURCES
+    circle.cpp
+    colliderbox.cpp
+    collisionmanager.cpp
+    input.cpp
+    mesh.cpp
+    pong.cpp
+    rectangle.cpp
+)
+
+add_library(engine ${SOURCES})
+
+target_link_libraries(engine
+    PUBLIC
+        render
+        gameobjects
```

Comment:
Do you understand the difference between public, private, and interface, and why each should be used?

Reply:
Kind of, as far as I understood, public is used here to link these libraries, then anything that links engine would not have to link render or gameobjects because it would also link against them automatically, while private would not. And interface is a weird middle case which includes properties or libraries for a target in a public manner but does not use them for itself

Reply:
Yep. So I think the easiest way to understand these is in terms of headers vs source files.

If something is PRIVATE it means the classes and `#include` for that dependency are only in the source files (.cpp)

If something is interface, it means the classes and `#include` for that dependency are only in the header files (.h)

PUBLIC is both headers and source.


In practice, you want to ensure that as many dependencies are PRIVATE as possible. This should be the default you go with, and too many dependencies being PUBLIC is a code smell that you have some wacky dependency structure going on. This affects code architecture, as well as compile times. Technically, you could always just link everything as PUBLIC, but that would mean if you change a source file in module A, ALL of your other code needs to be recompiled. If everything is PRIVATE, changes are self contained and don't leak elsewhere, improving compile times and encapsulation.

Reply:
All public dependencies have the ability to cascade to other modules, so if you have foo PUBLIC depends on bar, and bar PUBLIC depends on baz, baz will link foo.

You can see how quickly that gets out of hand. All your modules know too much about each other and lead to architectural nastiness.